langkit/templates/pkg_implementation_spec_ada.mako

```diff
@@ -558,6 +560,260 @@ private package ${ada_lib_name}.Implementation is
+   --     mapping addition, node's Self_Env assignment. This info is
+   --     tracked using the Named_Env_Descriptor record type below, often
+   --     abbreviated NED. Note that this tracking happens even when there is
+   --     no environment associated to the env name.
+   --
+   --  4. Unit reparsing can destroy existing environments and/or create new
+   --     ones. This means that, depending on their "ranking" using the
+   --     ordering predicate, environments can earn or lose precedence for a
+   --     given name.
+   --
+   --  5. When the precedence changes for a given name, we use the info
+   --     collected as per 3. to perform relocation: relevant environment
+   --     parent links are updated, symbol/node mappings are removed from the
+   --     env that lost precedence and added to the env that earned precedence,
+   --     etc.
+
+   --  Handling of environment names
+
+   type Env_Name_Record;
+   type Env_Name is access Env_Name_Record;
+   --  Name for an environment. Such names are non-empty sequences of symbols
+
+   type Env_Name_Internal_Array is array (Positive range <>) of Symbol_Type;
+   type Env_Name_Record (Size : Positive) is record
+      Symbols : Env_Name_Internal_Array (1 .. Size);
+
+      Hash : Hash_Type;
+      --  Precomputed hash for this name. No need to compute the hash more than
+      --  once.
+   end record;
+
+   procedure Destroy is new Ada.Unchecked_Deallocation
+     (Env_Name_Record, Env_Name);
+   function Equivalent (Left, Right : Env_Name) return Boolean
+   is (Left.Symbols = Right.Symbols);
+   function Hash (Name : Env_Name) return Hash_Type is (Name.Hash);
+
+   package Env_Name_Sets is new Ada.Containers.Hashed_Sets
+     (Element_Type        => Env_Name,
+      Hash                => Hash,
+      Equivalent_Elements => Equivalent);
+   --  Internalization table for named symbols (see the Create_Env_Name
+   --  function below).
+
+   --  Tables to populate lexical entries in named envs
+
+   package NED_Assoc_Maps is new Ada.Containers.Hashed_Maps
+     (Key_Type        => Symbol_Type,
+      Element_Type    => Internal_Map_Node_Vectors.Vector,
+      Hash            => Hash,
+      Equivalent_Keys => "=",
+      "="             => Internal_Map_Node_Vectors."=");
+   --  Symbol/lexical env entry mappings for a given named env descriptor.
+   --  Symbols are not unique in all mappings, so the lexical env entries are
+   --  stored in a vector.
+
+   procedure Add
+     (Self : in out NED_Assoc_Maps.Map;
+      Key  : Symbol_Type;
+      Node : AST_Envs.Internal_Map_Node);
+   --  Add a symbol/lexical env entry mapping in Self
+
+   procedure Remove
+     (Self : in out NED_Assoc_Maps.Map;
+      Key  : Symbol_Type;
+      Node : ${T.root_node.name});
+   --  Remove a symbol/lexical env entry mapping from Self
+
+   --  Global table for named environments
+
+   package Sorted_Env_Maps is new Ada.Containers.Ordered_Maps
```

Comment:
Calling it a set while it’s implemented as a map is also weird IMO. Also, I don’t think using a set will be a win, as it will complexify the uses:

* `Envs.First_Element` -> `Envs.First_Element.Env`
* `Envs.Delete (Node)` -> `Envs.Delete ((Node => Node, Env => Empty_Env));`

So let’s always refer to such objects as maps, no?

Reply:
OK!